File: frontend/js/views/rsu_grant.js
// RSU grant generator — cliff+monthly vesting, vest value (ordinary income),
// and sell-to-cover net-share delivery, via /calc/rsu-grant.

import { api } from '../api.js';
import { applyUiI18n, t } from '../i18n.js';
import { currentViewToken, viewIsCurrent } from '../app.js';
import { showToast } from '../toast.js';
import { debounce } from '../util.js';

const esc = (s) => String(s).replace(/[&<>]/g, (c) => ({ '&': '&amp;', '<': '&lt;', '>': '&gt;' }[c]));
const money = (n) => (n == null ? '—' : '$' + Number(n).toLocaleString(undefined, { minimumFractionDigits: 2, maximumFractionDigits: 2 }));
const num = (n) => (n == null ? '—' : Number(n).toLocaleString(undefined, { maximumFractionDigits: 0 }));
let LAST_DOC = null;

export async function renderRsuGrant(mount, _appState) {
    const tok = currentViewToken();
    if (!viewIsCurrent(tok)) return;
    mount.innerHTML = `
        <h1 class="view-title"><span data-i18n="view.rsu-grant.h1.title">// RSU GRANT</span></h1>
        <p class="muted small" data-i18n="view.rsu-grant.hint.intro">
            A restricted stock unit grant — a full-value award with no strike price and no exercise. On
            vesting, the fair-market value of the vested units is ordinary income; the employer withholds
            taxes by holding back shares (sell-to-cover) and delivers the net. Drafting aid, not legal/tax advice.
        </p>
        <div class="lpv-split">
        <div class="chart-panel">
            <h2 data-i18n="view.rsu-grant.h2.inputs">Grant terms</h2>
            <form id="rsu-form" class="inline-form">
                <label><span data-i18n="view.rsu-grant.label.state">State</span>
                    <input type="text" name="state" value="Delaware" required></label>
                <label><span data-i18n="view.rsu-grant.label.company">Company</span>
                    <input type="text" name="company_name" value=""></label>
                <label><span data-i18n="view.rsu-grant.label.grantee">Grantee</span>
                    <input type="text" name="grantee_name" value=""></label>
                <label><span data-i18n="view.rsu-grant.label.total">Total units</span>
                    <input type="number" step="1000" min="0" name="total_units" value="40000" required></label>
                <label><span data-i18n="view.rsu-grant.label.fmv">Current FMV ($)</span>
                    <input type="number" step="0.01" min="0" name="fmv_usd" value="10.00" required></label>
                <label><span data-i18n="view.rsu-grant.label.withholding">Tax withholding (%)</span>
                    <input type="number" step="0.1" min="0" name="withholding_pct" value="22"></label>
                <label><span data-i18n="view.rsu-grant.label.vesting">Vesting (months)</span>
                    <input type="number" step="1" min="1" name="vesting_months" value="48" required></label>
                <label><span data-i18n="view.rsu-grant.label.cliff">Cliff (months)</span>
                    <input type="number" step="1" min="0" name="cliff_months" value="12" required></label>
                <label><span data-i18n="view.rsu-grant.label.grant_date">Grant date</span>
                    <input type="date" name="grant_date" value="2024-01-01" required></label>
                <label><span data-i18n="view.rsu-grant.label.as_of">Valuation date</span>
                    <input type="date" name="as_of_date" value="2026-07-01" required></label>
                <label><span data-i18n="view.rsu-grant.label.statute">Statute citation (optional)</span>
                    <input type="text" name="statute_citation" value="" placeholder="${esc(t('view.rsu-grant.ph.statute'))}"></label>
            </form>
        </div>
        <div id="rsu-result" class="chart-panel lpv-preview"></div>
        </div>
    `;
    applyUiI18n(mount);

    const form = mount.querySelector('#rsu-form');
    const generate = async () => {
        const fd = new FormData(form);
        const body = {
            company_name: (fd.get('company_name') || '').trim(),
            grantee_name: (fd.get('grantee_name') || '').trim(),
            total_units: Number(fd.get('total_units')) || 0,
            fmv_usd: Number(fd.get('fmv_usd')) || 0,
            withholding_pct: Number(fd.get('withholding_pct')) || 0,
            vesting_months: Number(fd.get('vesting_months')) || 0,
            cliff_months: Number(fd.get('cliff_months')) || 0,
            grant_date: fd.get('grant_date'),
            as_of_date: fd.get('as_of_date'),
            state: (fd.get('state') || '').trim(),
            statute_citation: (fd.get('statute_citation') || '').trim(),
        };
        try {
            const doc = await api.calcRsuGrant(body);
            if (!viewIsCurrent(tok)) return;
            renderResult(mount, doc);
        } catch (err) {
            showToast(err.message || t('view.rsu-grant.toast.error'), { level: 'error' });
        }
    };
    const live = debounce(generate, 250);
    form.addEventListener('input', () => { live(); });
    form.addEventListener('submit', (e) => { e.preventDefault(); generate(); });
    generate();
}

function docToText(doc) {
    const lines = [doc.title.toUpperCase()];
    if (doc.statutory_citation) lines.push(doc.statutory_citation);
    lines.push('');
    for (const c of doc.clauses) lines.push(c.heading, c.body, '');
    return lines.join('\n');
}

function renderResult(mount, doc) {
    LAST_DOC = doc;
    const el = mount.querySelector('#rsu-result');
    el.innerHTML = `
        <div class="lpv-bar">
            <div class="cards">
                <div class="card pos"><div class="label" data-i18n="view.rsu-grant.card.vested">Vested units</div>
                    <div class="value">${num(doc.vested_units)}</div></div>
                <div class="card"><div class="label" data-i18n="view.rsu-grant.card.value">Vest value (income)</div>
                    <div class="value">${money(doc.vest_value_usd)}</div></div>
                <div class="card"><div class="label" data-i18n="view.rsu-grant.card.withheld">Tax withheld</div>
                    <div class="value">${money(doc.tax_withheld_usd)}</div></div>
                <div class="card"><div class="label" data-i18n="view.rsu-grant.card.shares_withheld">Shares withheld</div>
                    <div class="value">${num(doc.shares_withheld)}</div></div>
                <div class="card pos"><div class="label" data-i18n="view.rsu-grant.card.net">Net shares delivered</div>
                    <div class="value">${num(doc.net_shares_delivered)}</div></div>
            </div>
            <div class="btn-row-inline">
                <button class="btn btn-secondary" id="rsu-copy" type="button" data-i18n="view.rsu-grant.btn.copy">Copy</button>
                <button class="btn btn-secondary" id="rsu-download" type="button" data-i18n="view.rsu-grant.btn.download">Download .txt</button>
            </div>
        </div>
        <pre class="small">${esc(docToText(doc))}</pre>
    `;
    applyUiI18n(el);

    el.querySelector('#rsu-copy').addEventListener('click', async () => {
        try {
            await navigator.clipboard.writeText(docToText(LAST_DOC));
            showToast(t('view.rsu-grant.toast.copied'), { level: 'success' });
        } catch (e) {
            showToast(t('view.rsu-grant.toast.copy_failed', { err: e.message || String(e) }), { level: 'error' });
        }
    });
    el.querySelector('#rsu-download').addEventListener('click', () => {
        const blob = new Blob([docToText(LAST_DOC)], { type: 'text/plain' });
        const url = URL.createObjectURL(blob);
        const a = document.createElement('a');
        a.href = url;
        a.download = 'rsu-grant.txt';
        a.click();
        URL.revokeObjectURL(url);
    });
}

File: crates/traderview-core/src/rsu_grant.rs
//! Restricted stock unit (RSU) grant — a full-value equity award, distinct from
//! a stock option grant: there is no strike price, no exercise, and no AMT. On
//! vesting, the fair-market value of the vested units is ordinary income, and an
//! employer typically withholds taxes by holding back shares ("sell-to-cover"),
//! delivering the net. It reuses the shared cliff+monthly vesting schedule from
//! `option_grant` (no duplicated vesting math). Drafting aid, not legal/tax advice.

use crate::option_grant::{months_between, vested_shares};
use chrono::NaiveDate;
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Deserialize)]
pub struct RsuGrantInput {
    pub company_name: String,
    pub grantee_name: String,
    /// Total RSUs granted.
    pub total_units: f64,
    /// Current fair market value per share.
    pub fmv_usd: f64,
    /// Tax withholding rate on the vest value, percent (US supplemental wage
    /// default is 22%).
    #[serde(default = "default_withholding")]
    pub withholding_pct: f64,
    /// Total vesting term in months (e.g. 48 for a 4-year vest).
    pub vesting_months: u32,
    /// Cliff length in months before any vesting (e.g. 12).
    pub cliff_months: u32,
    pub grant_date: String,
    /// Date to value vesting against.
    pub as_of_date: String,
    pub state: String,
    #[serde(default)]
    pub statute_citation: String,
}

fn default_withholding() -> f64 {
    22.0
}

#[derive(Debug, Clone, Serialize, PartialEq)]
pub struct DocClause {
    pub heading: String,
    pub body: String,
}

#[derive(Debug, Clone, Serialize, PartialEq)]
pub struct RsuGrant {
    pub title: String,
    pub total_units: f64,
    pub months_elapsed: i64,
    pub vested_units: f64,
    pub unvested_units: f64,
    /// Ordinary income recognized on the vested units (fmv × vested).
    pub vest_value_usd: f64,
    /// Tax withheld at vest (withholding_pct × vest value).
    pub tax_withheld_usd: f64,
    /// Shares held back to fund the withholding (rounded up to cover it).
    pub shares_withheld: f64,
    /// Net shares delivered after sell-to-cover.
    pub net_shares_delivered: f64,
    pub statutory_citation: String,
    pub clauses: Vec<DocClause>,
}

fn cents(x: f64) -> f64 {
    (x * 100.0).round() / 100.0
}

fn money(v: f64) -> String {
    format!("${:.2}", v)
}

pub fn generate(i: &RsuGrantInput) -> RsuGrant {
    let grant = NaiveDate::parse_from_str(&i.grant_date, "%Y-%m-%d");
    let as_of = NaiveDate::parse_from_str(&i.as_of_date, "%Y-%m-%d");
    let elapsed = match (grant, as_of) {
        (Ok(g), Ok(a)) => months_between(g, a),
        _ => 0,
    };

    let vested = vested_shares(i.total_units, i.vesting_months, i.cliff_months, elapsed);
    let unvested = (i.total_units - vested).max(0.0);

    let vest_value = cents(i.fmv_usd * vested);
    let tax_withheld = cents(vest_value * i.withholding_pct / 100.0);
    // Whole shares held back, rounded up so the proceeds cover the withholding.
    let shares_withheld = if i.fmv_usd > 0.0 {
        (tax_withheld / i.fmv_usd).ceil()
    } else {
        0.0
    };
    let net_shares = (vested - shares_withheld).max(0.0);

    let citation = i.statute_citation.trim();
    let governing = if citation.is_empty() {
        format!("This grant is governed by the laws of the State of {}.", i.state)
    } else {
        format!("This grant is governed by the laws of the State of {} ({}).", i.state, citation)
    };

    let years = i.vesting_months as f64 / 12.0;
    let vesting_body = format!(
        "The RSUs vest over {} months ({:.2} years) with a {}-month cliff. No units vest before the cliff; thereafter units vest monthly in equal installments. As of {}, {} month(s) have elapsed and {:.0} of {:.0} units are vested ({:.0} unvested).",
        i.vesting_months, years, i.cliff_months, i.as_of_date, elapsed, vested, i.total_units, unvested
    );

    let settlement_body = format!(
        "On vesting, each unit settles into one share of common stock. The {} fair market value of the {:.0} vested units is ordinary income of {} in the year of vesting. At a {:.1}% withholding rate, {} of tax is withheld by holding back {:.0} shares (sell-to-cover), and {:.0} net shares are delivered. RSUs have no strike price and are not exercised.",
        money(i.fmv_usd), vested, money(vest_value), i.withholding_pct, money(tax_withheld), shares_withheld, net_shares
    );

    let clauses = vec![
        DocClause {
            heading: "Parties".into(),
            body: format!(
                "Company: {}\nGrantee: {}\nGrant date: {}",
                i.company_name, i.grantee_name, i.grant_date
            ),
        },
        DocClause {
            heading: "1. Grant".into(),
            body: format!(
                "The Company grants the Grantee {:.0} restricted stock units, each representing the right to receive one share of common stock on vesting, subject to the Company's equity plan and this agreement.",
                i.total_units
            ),
        },
        DocClause { heading: "2. Vesting".into(), body: vesting_body },
        DocClause {
            heading: "3. Forfeiture".into(),
            body: "Unvested units are forfeited automatically and without consideration on termination of service. Vesting does not continue after the service relationship ends except as expressly provided in the plan.".into(),
        },
        DocClause { heading: "4. Settlement & Tax Withholding".into(), body: settlement_body },
        DocClause {
            heading: "5. Transfer Restrictions".into(),
            body: "Unvested units are non-transferable. Shares delivered on settlement are subject to the Company's insider-trading policy and applicable securities-law transfer restrictions.".into(),
        },
        DocClause { heading: "6. Governing Law".into(), body: governing },
        DocClause {
            heading: "Signatures".into(),
            body: format!(
                "Company: ____________________  Date: __________\n{}\n\nGrantee: ____________________  Date: __________\n{}",
                i.company_name, i.grantee_name
            ),
        },
    ];

    RsuGrant {
        title: "Restricted Stock Unit (RSU) Grant".into(),
        total_units: i.total_units,
        months_elapsed: elapsed,
        vested_units: vested,
        unvested_units: unvested,
        vest_value_usd: vest_value,
        tax_withheld_usd: tax_withheld,
        shares_withheld,
        net_shares_delivered: net_shares,
        statutory_citation: citation.to_string(),
        clauses,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn close(a: f64, b: f64) -> bool {
        (a - b).abs() < 0.01
    }

    fn base() -> RsuGrantInput {
        RsuGrantInput {
            company_name: "Startup Inc".into(),
            grantee_name: "Jane Engineer".into(),
            total_units: 40_000.0,
            fmv_usd: 10.00,
            withholding_pct: 22.0,
            vesting_months: 48,
            cliff_months: 12,
            grant_date: "2024-01-01".into(),
            as_of_date: "2026-07-01".into(),
            state: "Delaware".into(),
            statute_citation: String::new(),
        }
    }

    #[test]
    fn vesting_30_months_in() {
        let g = generate(&base());
        assert_eq!(g.months_elapsed, 30);
        assert!(close(g.vested_units, 25_000.0));
        assert!(close(g.unvested_units, 15_000.0));
    }

    #[test]
    fn vest_value_and_sell_to_cover() {
        let g = generate(&base());
        assert!(close(g.vest_value_usd, 250_000.0));
        assert!(close(g.tax_withheld_usd, 55_000.0));
        // 55,000 / 10 = 5,500 shares held back; 25,000 - 5,500 = 19,500 delivered.
        assert!(close(g.shares_withheld, 5_500.0));
        assert!(close(g.net_shares_delivered, 19_500.0));
    }

    #[test]
    fn shares_withheld_round_up_to_cover() {
        // Odd FMV so tax/fmv is fractional and must round up.
        let g = generate(&RsuGrantInput { fmv_usd: 3.0, ..base() });
        // vested 25,000 × $3 = $75,000 income; 22% = $16,500 tax; 16500/3 = 5,500 exact.
        assert!(close(g.tax_withheld_usd, 16_500.0));
        assert!(close(g.shares_withheld, 5_500.0));
    }

    #[test]
    fn pre_cliff_zero() {
        let g = generate(&RsuGrantInput { as_of_date: "2024-07-01".into(), ..base() });
        assert_eq!(g.months_elapsed, 6);
        assert!(close(g.vested_units, 0.0));
        assert!(close(g.vest_value_usd, 0.0));
    }

    #[test]
    fn fully_vested_caps_at_total() {
        let g = generate(&RsuGrantInput { as_of_date: "2030-01-01".into(), ..base() });
        assert!(close(g.vested_units, 40_000.0));
        assert!(close(g.unvested_units, 0.0));
    }

    #[test]
    fn no_strike_or_exercise_language() {
        let c = generate(&base()).clauses.into_iter().find(|c| c.heading.contains("Settlement")).unwrap();
        assert!(c.body.contains("no strike price"));
        assert!(c.body.contains("ordinary income"));
    }

    #[test]
    fn zero_withholding_delivers_all_vested() {
        let g = generate(&RsuGrantInput { withholding_pct: 0.0, ..base() });
        assert!(close(g.shares_withheld, 0.0));
        assert!(close(g.net_shares_delivered, 25_000.0));
    }

    #[test]
    fn statute_citation_echoed() {
        let g = generate(&RsuGrantInput { statute_citation: "26 U.S.C. § 83".into(), ..base() });
        assert_eq!(g.statutory_citation, "26 U.S.C. § 83");
        assert!(g.clauses.iter().any(|c| c.body.contains("26 U.S.C. § 83")));
    }
}
